### PR TITLE
MemTableList::TrimHistory now use allocated bytes

### LINKED ISSUE #8371: rocksdb::Arena::AllocateNewBlock allocate memory grows without limit
> Note: Please use Issues only for bug reports. For questions, discussions, feature requests, etc. post to dev group: https://groups.google.com/forum/#!forum/rocksdb or https://www.facebook.com/groups/rocksdb.dev

### Expected behavior
rocksdb::Arena::AllocateNewBlock same with rocksdb.size-all-mem-tables, data_cf:815343456, log_cf:728651792
### Actual behavior
rocksdb::Arena::AllocateNewBlock have allocated about 17GB
rocksdb.size-all-mem-tables, data_cf:831672784, log_cf:870279456
rocksdb.estimate-table-readers-mem, data_cf:8345741792, log_cf:3513280
rocksdb.num-live-versions, data_cf:1, log_cf:1
rocksdb.estimate-pending-compaction-bytes, data_cf:612062269, log_cf:0
rocksdb.block-cache-usage, data_cf:1022153072, log_cf:1022153072
rocksdb.rocksdb.block-cache-pinned-usage, data_cf:0, log_cf:0

### Steps to reproduce the behavior
version:6.20.3 and 6.8.1
options：
[CFOptions "raft_log"]
  bottommost_compression=kDisableCompressionOption
  bottommost_compression_opts={max_dict_buffer_bytes=0;enabled=false;parallel_threads=1;zstd_max_train_bytes=0;strategy=0;max_dict_bytes=0;level=32767;window_bits=-14;}
  sample_for_compression=0
  blob_file_size=268435456
  compaction_options_universal={allow_trivial_move=false;stop_style=kCompactionStopStyleTotalSize;compression_size_percent=-1;max_size_amplification_percent=200;max_merge_width=4294967295;min_merge_width=2;size_ratio=1;}
  compaction_options_fifo={allow_compaction=false;max_table_files_size=1073741824;}
  prefix_extractor=rocksdb.FixedPrefix.9
  max_bytes_for_level_multiplier_additional=1:{1}:{1}:{1}:{1}:{1}:{1}
  max_bytes_for_level_base=1073741824
  memtable_whole_key_filtering=false
  memtable_prefix_bloom_size_ratio=0.000000
  enable_blob_files=false
  target_file_size_base=134217728
  memtable_huge_page_size=0
  max_successive_merges=0
  inplace_update_num_locks=10000
  max_sequential_skip_in_iterations=8
  arena_block_size=16777216
  target_file_size_multiplier=1
  max_write_buffer_number=6
  write_buffer_size=134217728
  blob_compression_type=kNoCompression
  compression=kSnappyCompression
  level0_stop_writes_trigger=160
  level0_slowdown_writes_trigger=10
  level0_file_num_compaction_trigger=5
  ttl=2592000
  max_compaction_bytes=3355443200
  blob_garbage_collection_age_cutoff=0.250000
  compression_opts={max_dict_buffer_bytes=0;enabled=false;parallel_threads=1;zstd_max_train_bytes=0;strategy=0;max_dict_bytes=0;level=32767;window_bits=-14;}
  enable_blob_garbage_collection=false
  soft_pending_compaction_bytes_limit=68719476736
  paranoid_file_checks=false
  periodic_compaction_seconds=0
  check_flush_compaction_key_order=true
  min_blob_size=0
  hard_pending_compaction_bytes_limit=274877906944
  disable_auto_compactions=false
  max_bytes_for_level_multiplier=10.000000
  report_bg_io_stats=false
  compaction_pri=kOldestLargestSeqFirst
  compaction_filter_factory=nullptr
  comparator=leveldb.BytewiseComparator
  bloom_locality=0
  compaction_style=kCompactionStyleLevel
  min_write_buffer_number_to_merge=2
  max_write_buffer_size_to_maintain=805306368
  max_write_buffer_number_to_maintain=0
  merge_operator=nullptr
  memtable_factory=SkipListFactory

[CFOptions "data"]
  bottommost_compression=kZSTD
  bottommost_compression_opts={max_dict_buffer_bytes=0;enabled=true;parallel_threads=1;zstd_max_train_bytes=262144;strategy=0;max_dict_bytes=16384;level=32767;window_bits=-14;}
  sample_for_compression=0
  blob_file_size=268435456
  compaction_options_universal={allow_trivial_move=false;stop_style=kCompactionStopStyleTotalSize;compression_size_percent=-1;max_size_amplification_percent=200;max_merge_width=4294967295;min_merge_width=2;size_ratio=1;}
  compaction_options_fifo={allow_compaction=false;max_table_files_size=1073741824;}
  prefix_extractor=rocksdb.FixedPrefix.16
  max_bytes_for_level_multiplier_additional=1:{1}:{1}:{1}:{1}:{1}:{1}
  max_bytes_for_level_base=1073741824
  memtable_whole_key_filtering=false
  memtable_prefix_bloom_size_ratio=0.000000
  enable_blob_files=false
  target_file_size_base=134217728
  memtable_huge_page_size=0
  max_successive_merges=0
  inplace_update_num_locks=10000
  max_sequential_skip_in_iterations=8
  arena_block_size=16777216
  target_file_size_multiplier=1
  max_write_buffer_number=6
  write_buffer_size=134217728
  blob_compression_type=kNoCompression
  compression=kSnappyCompression
  level0_stop_writes_trigger=160
  level0_slowdown_writes_trigger=10
  level0_file_num_compaction_trigger=5
  ttl=2592000
  max_compaction_bytes=3355443200
  blob_garbage_collection_age_cutoff=0.250000
  compression_opts={max_dict_buffer_bytes=0;enabled=false;parallel_threads=1;zstd_max_train_bytes=0;strategy=0;max_dict_bytes=0;level=32767;window_bits=-14;}
  enable_blob_garbage_collection=false
  soft_pending_compaction_bytes_limit=68719476736
  paranoid_file_checks=false
  periodic_compaction_seconds=2592000
  check_flush_compaction_key_order=true
  min_blob_size=0
  hard_pending_compaction_bytes_limit=549755813888
  disable_auto_compactions=false
  max_bytes_for_level_multiplier=5.000000
  report_bg_io_stats=false
  compaction_pri=kByCompensatedSize
  compaction_filter_factory=nullptr
  comparator=leveldb.BytewiseComparator
  bloom_locality=0
  compaction_style=kCompactionStyleLevel
  min_write_buffer_number_to_merge=2
  max_write_buffer_size_to_maintain=805306368
  max_write_buffer_number_to_maintain=0
  merge_operator=nullptr
  memtable_factory=SkipListFactory
  memtable_insert_with_hint_prefix_extractor=nullptr
  compression_per_level=kNoCompression:{kNoCompression}:{kLZ4Compression}:{kLZ4Compression}:{kLZ4Compression}:{kLZ4Compression}:{kLZ4Compression}
  num_levels=7
  force_consistency_checks=true
  optimize_filters_for_hits=false
  compaction_filter=SplitCompactionFilter
  level_compaction_dynamic_level_bytes=true
  inplace_update_support=false
  table_factory=BlockBasedTable

[TableOptions/BlockBasedTable "data"]
  pin_top_level_index_and_filter=true
  block_align=false
  read_amp_bytes_per_bit=0
  verify_compression=false
  enable_index_compression=true
  whole_key_filtering=true
  max_auto_readahead_size=262144
  optimize_filters_for_memory=false
  index_block_restart_interval=1
  block_restart_interval=16
  block_size=65536
  format_version=4
  partition_filters=false
  block_size_deviation=10
  no_block_cache=false
  checksum=kCRC32c
  data_block_hash_table_util_ratio=0.750000
  index_shortening=kShortenSeparators
  data_block_index_type=kDataBlockBinaryAndHash
  hash_index_allow_collision=true
  filter_policy=rocksdb.BuiltinBloomFilter
  metadata_block_size=4096
  index_type=kBinarySearch
  metadata_cache_options={unpartitioned_pinning=kFallback;partition_pinning=kFallback;top_level_index_pinning=kFallback;}
  pin_l0_filter_and_index_blocks_in_cache=false
  cache_index_and_filter_blocks_with_high_priority=true
  cache_index_and_filter_blocks=false
  flush_block_policy_factory=FlushBlockBySizePolicyFactory

It grows faster when file ingest in rocksdb
image

### PR DESCRIPTION
**Pull request by lgqss**
Fix a bug when both max_write_buffer_size_to_maintain and max_write_buffer_number_to_maintain are 0. 
The bug was introduced in 6.5.0 and  #5022.
Fix #8371

### CODE REVIEW

**Comment by ajkr** [comment]
Thanks I agree there is a problem from the log lines. Looking at:

```
max_write_buffer_size_to_maintain_:805306368 max_write_buffer_number_to_maintain_:0 memlist_history_.size:88 ApproximateMemoryUsageExcludingLast:6908072
```

It is using `arena_block_size * memlist_history_.size() = 16MB * 88 = 1408MB` memory to maintain < 7MB of write history. And that'll keep growing since write history limit is `max_write_buffer_size_to_maintain = 768MB`.

One thing that should help is upgrading to 6.21 or higher. #7907 reduced the max arena_block_size to 1MB.

**Comment by ajkr** [comment]
I wonder if it would work for us to rewrite memtables with large wastage during flush before they become history. For example when `ApproximateMemoryUsage() << 1MB` we could create a new memtable for the history with allocation size `ApproximateMemoryUsage()` and copy into that.

But it's a non-trivial change. If you are able to avoid tiny memtables and/or disproportionately large arena_block_size it might be good enough. You could also try max_write_buffer_number_to_maintain as you mentioned, with the downside that the actual amount of write history will be less predictable.

**Comment by lgqss** [comment]
Should we use arena_block_size instead of ApproximateMemoryUsage() if ApproximateMemoryUsage() < arena_block_size.

**Comment by ajkr** [comment]
> Could we use arena_block_size instead of ApproximateMemoryUsage() if ApproximateMemoryUsage() < arena_block_size? When ApproximateMemoryUsage() >= arena_block_size, same the as current code. When ApproximateMemoryUsage() < arena_block_size, we can keep (max_write_buffer_number_to_maintain/arena_block_size) memtables . It's enough.

I think we should consider allocated bytes, yes. But we should list it under "Behavior Changes" in HISTORY.md. Also can we always use allocated bytes, for example, by introducing a MemTable::MemoryAllocatedBytes() function and using that instead? We could use this formula to implement it. https://github.com/facebook/rocksdb/blob/12117b26a359dc4e9649994b96991acb0720b3ff/db/memtable.cc#L164-L166

**Comment by lgqss** [comment]
> Also can we always use allocated bytes
I think it's ok in 1MB arena_block_size. 
If users set a large value(e.g. 16MB) , I don't quite know what the impact is.

File: db/memtable_list.cc [hunk 1]
```diff
@@ -259,8 +259,8 @@ SequenceNumber MemTableListVersion::GetEarliestSequenceNumber(
 void MemTableListVersion::Add(MemTable* m, autovector<MemTable*>* to_delete) {
   assert(refs_ == 1);  // only when refs_ == 1 is MemTableListVersion mutable
   AddMemTable(m);
-
-  TrimHistory(to_delete, m->ApproximateMemoryUsage());
+  // m->ApproximateMemoryUsage() is added in ApproximateMemoryUsageExcludingLast
```
Comment (ajkr) [inline db/memtable_list.cc:262]: Good point.

**Review by ajkr — CHANGES_REQUESTED** [review]

File: HISTORY.md [hunk 2]
```diff
@@ -1,4 +1,8 @@
 # Rocksdb Change Log
+## Unreleased
+### Behavior Changes
+* MemTableListVersion::ApproximateMemoryUsageExcludingLast now use allocated bytes(arena_block_size) when history memtable's memory_usgage < arena_block_size. This change only affects immutable memtables with few data in trimming logic (max_write_buffer_size_to_maintain > 0, defalut in trasactiondb). Fix #8371.
```
[the remark this answers is not in the document]
Comment (lgqss) [reply HISTORY.md]: I think it's confused for others if we just use allocated bytes in first arena block.
So I think we need always use allocated bytes as you suggested

[the remark this answers is not in the document]
Comment (lgqss) [reply HISTORY.md]: I think it's confused for others if we just use allocated bytes in first arena block.
So I think we need always use allocated bytes as you suggested.

**Review by ajkr — APPROVED** [review]
Looks great, thank you very much!

**Pull request closed without merging**